cf-check/backup.c

```diff
@@ -0,0 +1,198 @@
+#include <platform.h>
+#include <backup.h>
+
+#if defined (__MINGW32__)
+
+int backup_main(int argc, char **argv)
+{
+    printf("backup not supported on Windows\n");
+    return 1;
+}
+
+#elif ! defined (LMDB)
+
+int backup_main(int argc, char **argv)
+{
+    printf("backup only implemented for LMDB.\n");
+    return 1;
+}
+
+#else
+
+#include <stdio.h>
+
+bool copy_file(const char *src, const char *dst)
+{
+    assert(src != NULL);
+    assert(dst != NULL);
+
+    printf("Copying: '%s' -> '%s'\n", src, dst);
+    FILE *in = fopen(src, "r");
+    if (in == NULL)
+    {
+        return false;
+    }
+    FILE *out = fopen(dst, "w");
+    if (out == NULL)
+    {
+        fclose(in);
+        return false;
+    }
+    size_t read = 0;
+    size_t written = 0;
+    bool ret = true;
+    do
+    {
+        const size_t bufsize = 1024;
+        char buf[1024] = { 0 };
+
+        read = fread(buf, sizeof(char), bufsize, in);
+
+        written = fwrite(buf, sizeof(char), read, out);
+        while ((!ferror(out)) && (written < read))
+        {
+            written += fwrite(buf + written, sizeof(char), read - written, out);
+        }
+
+        if (read != written || ferror(in) || ferror(out))
+        {
+            break;
+        }
+        if (feof(in))
+        {
+            read = 0;
+        }
+    } while (read > 0);
+
+    if (ferror(in))
+    {
+        printf("Error encountered while reading '%s'\n", src);
+        ret = false;
+    }
+    else if (ferror(out))
+    {
+        printf("Error encountered while writing '%s'\n", dst);
+        ret = false;
+    }
+    else if (read != 0 && read != written)
+    {
+        printf("Did not copy the whole file\n");
+        ret = false;
+    }
+
+    fclose(in);
+    fclose(out);
```

Comment:
check for errors here at fclose() since that is when the buffered FILE * is sure to be flushed?

Reply:
Looks like I should check errno here, yes:

```
RETURN VALUES
     Upon successful completion 0 is returned.  Otherwise, EOF is returned and the global variable errno
     is set to indicate the error.  In either case no further access to the stream is possible.

ERRORS
     The fclose() function may also fail and set errno for any of the errors specified for the routines
     close(2) or fflush(3).
```